### PR TITLE
Replace rapidjson by json for modern c++

### LINKED ISSUE #38: Consider using Json for Modern C++ in place of rapidJson
See https://github.com/nlohmann/json
Conan recipe: https://github.com/vthiery/conan-jsonformoderncpp

### PR DESCRIPTION
**Pull request by dominikwerder**
Removes RapidJSON from the project and introduces JSON for modern C++ instead.
Also cleanup and LLVMification on the way.
Closes #38

### CODE REVIEW

File: src/Main.cpp [hunk 1]
```diff
@@ -45,9 +43,9 @@ Main::Main(MainOpt &opt)
   finfo = std::shared_ptr<ForwarderInfo>(new ForwarderInfo(this));
```
Comment (matthew-d-jones) [inline src/Main.cpp]: Use `make_shared`

File: src/Main.cpp [hunk 2]
```diff
@@ -45,9 +43,9 @@ Main::Main(MainOpt &opt)
...
   finfo->teamid = main_opt.teamid;
 
-  for (int i1 = 0; i1 < opt.conversion_threads; ++i1) {
+  for (size_t i1 = 0; i1 < opt.ConversionThreads; ++i1) {
     conversion_workers.emplace_back(new ConversionWorker(
-        &conversion_scheduler, opt.conversion_worker_queue_size));
+        &conversion_scheduler, opt.ConversionWorkerQueueSize));
```
Comment (matthew-d-jones) [inline src/Main.cpp]: `opt.ConversionWorkerQueueSize` is cast to `uint32_t` from `size_t`, use `static_cast` if intentional.

File: src/Main.cpp [hunk 3]
```diff
@@ -110,45 +109,42 @@ class ConfigCB : public Config::Callback {
...
-      return;
+}
+
+void ConfigCB::handleParsedJSON(nlohmann::json const &Document) {
+  using std::string;
+  using nlohmann::json;
+  if (auto x = find<string>("cmd", Document)) {
+    auto Command = x.inner();
+    if (Command == "add") {
+      if (auto x = find<json>("streams", Document)) {
```
Comment (matthew-d-jones) [inline src/Main.cpp]: I'm not keen on using `x` here at all, but it should definitely be avoided when there is another `x` in the upper scope.

File: src/Main.cpp [hunk 4]
```diff
@@ -110,45 +109,42 @@ class ConfigCB : public Config::Callback {
...
+            main.mappingAdd(Stream);
+          }
+        }
+      }
     }
-    if (m2->value.IsArray()) {
-      for (auto &x : m2->value.GetArray()) {
-        main.mapping_add(x);
+    if (Command == "stop_channel") {
+      if (auto x = find<string>("channel", Document)) {
```
Comment (matthew-d-jones) [inline src/Main.cpp]: Same here, `x` in upper scope.

File: src/Main.cpp [hunk 5]
```diff
@@ -242,23 +238,18 @@ void Main::forward_epics_to_kafka() {
...
-  jd.SetObject();
-  Value j_streams;
-  j_streams.SetArray();
-  for (auto &stream : streams.get_streams()) {
-    j_streams.PushBack(Value().CopyFrom(stream->status_json(), a), a);
+  using nlohmann::json;
+  auto Status = json::object();
+  auto Streams = json::array();
+  for (auto const &Stream : streams.get_streams()) {
+    // j_streams.PushBack(Value().CopyFrom(stream->status_json(), a), a);
```
Comment (matthew-d-jones) [inline src/Main.cpp]: Commented line can be deleted.

File: src/Main.cpp [hunk 6]
```diff
@@ -318,91 +309,96 @@ void Main::report_stats(int dt) {
...
+    uri::URI TopicURI;
+    if (!URI.host.empty()) {
+      TopicURI.host = URI.host;
+    }
+    if (URI.port != 0) {
+      TopicURI.port = URI.port;
+    }
+    TopicURI.parse(Topic);
+    Converter::sptr ConverterShared;
+    if (ConverterName.size() > 0) {
```
Comment (matthew-d-jones) [inline src/Main.cpp]: Use `!ConverterName.empty()`

File: src/MainOpt.cpp [hunk 7]
```diff
@@ -125,98 +103,50 @@ void MainOpt::parse_document(const std::string &filepath) {
...
-}
-
-void MainOpt::find_broker() {
-  auto itr = json->FindMember("broker");
-  if (itr != json->MemberEnd()) {
-    if (itr->value.IsString()) {
-      set_broker(itr->value.GetString());
-    }
-  }
+  return Settings;
```
Comment (matthew-d-jones) [inline src/MainOpt.cpp]: We can reach this point without `Settings` being initialised, would that cause problems or does it fall back on a default somewhere?

File: src/MainOpt.h [hunk 8]
```diff
@@ -20,8 +20,8 @@ struct MainOpt {
   std::string KafkaGELFAddress = "";
   std::string GraylogLoggerAddress = "";
   std::string InfluxURI = "";
-  int conversion_threads = 1;
-  uint32_t conversion_worker_queue_size = 1024;
+  size_t ConversionThreads = 1;
+  size_t ConversionWorkerQueueSize = 1024;
   std::string LogFilename;
   std::string ConfigurationFile;
   int forwarder_ix = 0;
```
Comment (matthew-d-jones) [inline src/MainOpt.h]: I don't think `forwarder_ix` and `write_per_message` are used.

File: src/MainOpt.h [hunk 9]
```diff
@@ -30,22 +30,23 @@ struct MainOpt {
   uint64_t teamid = 0;
   std::vector<char> Hostname;
   FlatBufs::SchemaRegistry schema_registry;
-  std::shared_ptr<rapidjson::Document> json;
+  nlohmann::json JSONConfiguration;
   int parse_json_file(std::string ConfigurationFile);
   KafkaW::BrokerSettings broker_opt;
   void init_logger();
-  void find_broker();
   void find_broker_config(uri::URI &property);
```
Comment (matthew-d-jones) [inline src/MainOpt.h]: `find_broker_config` isn't implemented.

File: src/json.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <nlohmann/json.hpp>
+#include <stdexcept>
+#include <string>
+
+template <typename T> class JsonMaybe {
+public:
+  JsonMaybe() {}
```
Comment (matthew-d-jones) [inline src/json.h]: Use `= default`

File: src/json.h [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  JsonMaybe(T inner) : inner_(inner), found_(true) {}
```
Comment (matthew-d-jones) [inline src/json.h]: This constructor should be marked `explicit`

File: src/MainOpt.cpp (hunk 7 above)
Comment (dominikwerder) [reply src/MainOpt.cpp]: It's not a problem. `Settings` is initialized in the first line of the function.

Comment (matthew-d-jones) [reply src/MainOpt.cpp]: It's declared there, not initialised.

Comment (dominikwerder) [reply src/MainOpt.cpp]: It's both declared and initialised, meaning the constructor is run and the object is in a valid state.

Comment (matthew-d-jones) [reply src/MainOpt.cpp]: Ah true, it's fine in this case, just looks wrong to me without `{}` somehow.

File: src/CMakeLists.txt [hunk 12]
```diff
@@ -18,7 +17,6 @@ set(path_include_common
 ${FMT_INCLUDE_DIR}
 ${RDKAFKA_INCLUDE_DIR}
 ${FLATBUFFERS_INCLUDE_DIR}
-${RAPIDJSON_INCLUDE_DIR}
 ${CURL_INCLUDE_DIRS}
```
Comment (mattclarke) [inline src/CMakeLists.txt]: Do we need JSONFMC++ adding to this?

File: src/Main.cpp [hunk 13]
```diff
@@ -42,12 +40,12 @@ Main::Main(MainOpt &opt)
...
-  finfo = std::shared_ptr<ForwarderInfo>(new ForwarderInfo(this));
+  finfo = std::make_shared<ForwarderInfo>(this);
   finfo->teamid = main_opt.teamid;
 
-  for (int i1 = 0; i1 < opt.conversion_threads; ++i1) {
+  for (size_t i1 = 0; i1 < opt.ConversionThreads; ++i1) {
     conversion_workers.emplace_back(new ConversionWorker(
-        &conversion_scheduler, opt.conversion_worker_queue_size));
+        &conversion_scheduler, static_cast<uint32_t>(opt.ConversionWorkerQueueSize)));
   }
```
Comment (mattclarke) [inline src/Main.cpp]: Use make_unique to avoid (unlikely) memory leaks

File: src/Main.cpp [hunk 14]
```diff
@@ -110,45 +109,42 @@ class ConfigCB : public Config::Callback {
...
-    LOG(3, "Command does not look like valid json");
-    return;
-  }
-  auto m1 = j0.FindMember("cmd");
-  if (m1 == j0.MemberEnd()) {
-    return;
-  }
-  if (!m1->value.IsString()) {
-    return;
+  try {
```
Comment (mattclarke) [inline src/Main.cpp]: c_str not needed

File: src/Main.cpp [hunk 15]
```diff
@@ -110,45 +109,42 @@ class ConfigCB : public Config::Callback {
...
-      return;
+}
+
+void ConfigCB::handleParsedJSON(nlohmann::json const &Document) {
+  using std::string;
+  using nlohmann::json;
+  if (auto CommandMaybe = find<string>("cmd", Document)) {
+    auto Command = CommandMaybe.inner();
+    if (Command == "add") {
+      if (auto StreamsMaybe = find<json>("streams", Document)) {
```
Comment (mattclarke) [inline src/Main.cpp]: separate function for add

File: src/Main.cpp [hunk 16]
```diff
@@ -110,45 +109,42 @@ class ConfigCB : public Config::Callback {
...
-  }
-  if (cmd == "stop_channel") {
-    auto channel = get_string(&j0, "channel");
-    if (channel.size() > 0) {
-      main.streams.channel_stop(channel);
+    if (Command == "stop_all") {
+      main.streams.streams_clear();
+    }
+    if (Command == "exit") {
+      main.stopForwarding();
```
Comment (mattclarke) [inline src/Main.cpp]: Log unrecognised commands

File: src/Main.cpp [hunk 17]
```diff
@@ -318,91 +308,96 @@ void Main::report_stats(int dt) {
...
-          if (!conv) {
-            conv =
-                Converter::create(main_opt.schema_registry, schema, main_opt);
-            converters[cname] = std::weak_ptr<Converter>(conv);
-          }
-        } else {
-          conv = Converter::create(main_opt.schema_registry, schema, main_opt);
-          converters[cname] = std::weak_ptr<Converter>(conv);
+  auto Stream = streams.back();
+  auto PushConverterToStream = [this](
```
Comment (mattclarke) [inline src/Main.cpp]: Too big for a lambda

File: src/MainOpt.cpp [hunk 18]
```diff
@@ -64,59 +59,42 @@ int MainOpt::parse_json_file(std::string ConfigurationFile) {
     return -1;
   }
```
Comment (mattclarke) [inline src/MainOpt.cpp]: Exception?

File: src/CMakeLists.txt (hunk 12 above)
Comment (dominikwerder) [reply src/CMakeLists.txt]: As far as I can see, there is no Find module for cmake.

**Pull request merged**